packages/adp-tooling/src/preview/client/dialogs/fragment.ts

```diff
@@ -0,0 +1,436 @@
+import { ValueState } from 'sap/ui/core/library';
+
+/** sap.ui.model */
+import JSONModel from 'sap/ui/model/json/JSONModel';
+
+/** sap.ui.rta */
+import CommandFactory from 'sap/ui/rta/command/CommandFactory';
+import type RuntimeAuthoring from 'sap/ui/rta/RuntimeAuthoring';
+
+/** sap.ui.base */
+import type Event from 'sap/ui/base/Event';
+import type EventProvider from 'sap/ui/base/EventProvider';
+import type ManagedObject from 'sap/ui/base/ManagedObject';
+import type ManagedObjectMetadata from 'sap/ui/base/ManagedObjectMetadata';
+
+/** sap.ui.dt */
+import OverlayRegistry from 'sap/ui/dt/OverlayRegistry';
+import type ElementOverlay from 'sap/ui/dt/ElementOverlay';
+import type ContextMenu from 'sap/ui/dt/plugin/ContextMenu';
+
+/** sap.ui.fl */
+import type { Layer } from 'sap/ui/fl';
+
+import ControlUtils from '../control-utils';
+import type { FragmentsResponse } from '../api-handler';
+import type { BuiltRuntimeControl } from '../control-utils';
+import { getFragments, getManifestAppdescr, writeFragment } from '../api-handler';
+
+interface CreateFragmentProps {
+    fragmentName: string;
+    index: string | number;
+    targetAggregation: string;
+}
+
+export interface ManifestAppdescr {
+    fileName: string;
+    layer: Layer;
+    fileType: string;
+    reference: string;
+    id: string;
+    namespace: string;
+    version: string;
+    content: object[];
+}
+
+interface DialogueData {
+    runtimeControl: ManagedObject;
+    control: BuiltRuntimeControl;
+}
+
+type ExtendedEventProvider = EventProvider & {
+    setEnabled: (v: boolean) => {};
+    getValue: () => string;
+    getSelectedItem: () => {
+        getTitle: () => string;
+        getText: () => string;
+        getCustomData: () => {
+            [key: string]: Function;
+        }[];
+    };
+    getSelectedKey: () => string;
+    setValueState: (state: ValueState) => void;
+    setValueStateText: (text: string) => void;
+    setVisible: (bool: boolean) => void;
+};
+
+/**
+ * Handles creation of the dialog, fills it with data
+ */
+export default class FragmentDialog {
+    /**
+     * Dialog instance
+     */
+    private dialog: Dialog;
+    /**
+     * JSON Model for the dialog
+     */
+    private model: JSONModel;
+    /**
+     * Runtime control managed object
+     */
+    private runtimeControl: ManagedObject;
+
+    /**
+     * @param rta Runtime Authoring
+     */
+    constructor(private rta: RuntimeAuthoring) {}
+
+    /**
+     * @description Initilizes "Add XML Fragment" functionality and adds a new item to the context menu
+     * @param contextMenu Context Menu from RTA
+     */
+    public init(contextMenu: ContextMenu) {
+        // We need this in order to keep the reference to the class
+        // we cannot use this keyword, because it is overshadowed by function scope
+        const that = this;
+
+        /**
+         * Controller for the action in the Dialog
+         */
+        const dummyController = {
```

Comment:
The name should already indicate that this is not the best way of defining the controller logic for you dialog. Please consider moving this into a separate file e.g. `addFragmentDialog.controller.ts` file.